Sofa/Component/LinearSolver/Iterative/src/sofa/component/linearsolver/iterative/MatrixLinearSolver.inl

```diff
@@ -240,31 +245,93 @@ void MatrixLinearSolver<Matrix,Vector>::invertSystem()
 template<class Matrix, class Vector>
 bool MatrixLinearSolver<Matrix,Vector>::addJMInvJtLocal(Matrix * /*M*/,ResMatrixType * result,const JMatrixType * J, SReal fact)
 {
-    for (typename JMatrixType::Index row=0; row<J->rowSize(); row++)
+    auto* taskScheduler = sofa::simulation::TaskScheduler::getInstance();
+
+    if( d_multithread.getValue() )
     {
-        // STEP 1 : put each line of matrix Jt in the right hand term of the system
-        for (typename JMatrixType::Index i=0; i<J->colSize(); i++) linearSystem.systemRHVector->set(i, J->element(row, i)); // linearSystem.systemMatrix->rowSize()
+        if (taskScheduler->getThreadCount() < 1)
+        {
+            taskScheduler->init(0);
+            msg_info() << "Task scheduler initialized on " << taskScheduler->getThreadCount() << " threads";
+        }
+        else
+        {
+            msg_info() << "Task scheduler already initialized on " << taskScheduler->getThreadCount() << " threads";
+        }
+    }
+    else
+    {
+        if (taskScheduler->getThreadCount() != 1) taskScheduler->init(1);
+    }
 
-        // STEP 2 : solve the system :
-        solveSystem();
+    
+    std::vector< solverTask<Matrix,Vector> > taskList;
+    
+    if( d_multithread.getValue() )  taskList.reserve( J->rowSize() );
+    
+    sofa::simulation::CpuTask::Status status;
+    
+    
+        
+    sofa::type::vector< Vector  > listRH(J->rowSize()) ; // columns of Jt
+    for(int i=0;i<J->rowSize();i++) listRH[i].resize(J->colSize());
+
+    sofa::type::vector< Vector  > listLH(J->rowSize()); // columns of MinvJt
+    for(int i=0;i<J->rowSize();i++) listLH[i].resize(J->colSize());
+
+    if (this->linearSystem.needInvert)
+        {
+            this->invert(*(this->linearSystem.systemMatrix));
+            this->linearSystem.needInvert = false;
+        }
 
-        // STEP 3 : project the result using matrix J
+    //STEPS 1&2: Copy Jt and compute Minv*Jt
+    {
+        sofa::helper::ScopedAdvancedTimer solveTimer("solve");
+        // one task per column of Jt
+        for (typename JMatrixType::Index col=0; col<J->rowSize(); col++)
+        {   
+            if( d_multithread.getValue() ) 
+            {
+                taskList.emplace_back(col , &listRH[col] , &listLH[col], &status , J , this  );
+                taskScheduler->addTask( &(taskList.back()) ); 
+            }
+            else
+            {
+                solverTask<Matrix,Vector> task(col , &listRH[col] , &listLH[col], &status , J , this  );
+                task.run();
+            }
+        }
+        taskScheduler->workUntilDone(&status);
+    }
+
+    // STEP 3 : compute the matricial product L*MinvJt
+    
+    sofa::linearalgebra::FullMatrix<double> product(J->rowSize(),J->rowSize());
```

Comment:
I am surprised that you have a specific type for the product. Why isn't it a `Vector`? What was used before?

Reply:
It is a temporary matrix used to store the result of the product.
There was no need for it before because the result was written on the Res matrix as soon as it was computed.